### DISCUSSION TITLE
Custom user registration

### DISCUSSION THREAD

**Discussion by dewebeloper** [OrchardCMS/OrchardCore#9264] [General]
What is correct way for custom user registration.
I have to extend User with FirstName, LastName and a few more fields and to create registration form where user can enter all info at once.
I tried to add fields as CustomUserSettings and create workflow (Register User task) for registration but currently I don't see way how to write down my additional fields to user.
Is this a good approach or would be better to rewrite registration page and is there any guide for that?

Thanks

**Comment by deanmarcussen** [comment]
Suggest rewriting registration pages. We are supporting editing custom fields from the admin, but not for the front end.

**Reply by jptissot** [reply]
I never tried, but a Form + Workflow could also work. I think we have Workflow tasks to create the User.

**Reply by Skrypt** [reply]
We don't have a Workflow task to create a user but we have event listeners on when a user is registered.
For a custom user registration form I basically created a module which uses a modified copy of the OrchardCore.Users RegistrationController and views.

So, I'm using the Workflow event for sending a email to an admin when a user is registered.

**Comment by dewebeloper** [comment]
@jptissot Yes, Form + Workflow works but doesn't save CustomUserProperties fields from form.

@Skrypt that is possibility, but again that is custom development. I want do skip custom development.

@deanmarcussen is there is way to create ContentType and then in Workflow somehow link registered user with ContentType?

**Reply by jptissot** [reply]
@dewebeloper look at what I have done here https://github.com/StatCan/StatCan.OrchardCore/blob/master/src/Modules/StatCan.OrchardCore.Scripting/UserGlobalMethodsProvider.cs . It allows you to edit the user's properties from a script task in a workflow.

**Reply by dewebeloper** [reply]
Thanks, will do

**Reply by jptissot** [reply]
I also have an example form in there to manage custom user settings. Just clone the solution, run it and try to use the User Profile form recipe

**Comment by yassinsayed** [comment]
this is my example with custom registration 
Clone and update the registration controller with below code add the addition attribute in view 

                if (iUser is User user)
               {
                    // Create UserProfile to website
                    IEnumerable<ContentTypeDefinition> GetContentTypeDefinitions = _contentDefinitionManager.ListTypeDefinitions()
                        .Where(x => x.GetSettings<ContentTypeSettings>().Stereotype == "CustomUserSettings");

                    if (GetContentTypeDefinitions != null)
                    {
                        var def = _contentDefinitionManager.GetTypeDefinition("UserProfile");
                        var contentItem = await _contentManager.NewAsync(def.Name);
                        if (contentItem != null)
                        {
                            if (contentItem.Content.UserProfile.ContainsKey("FirstName"))
                                contentItem.Content.UserProfile.FirstName.Text = model.FirstName.ToString();
                            if (contentItem.Content.UserProfile.ContainsKey("MiddleName"))
                                contentItem.Content.UserProfile.MiddleName.Text = model.MiddleName.ToString();
                            if (contentItem.Content.UserProfile.ContainsKey("LastName"))
                                contentItem.Content.UserProfile.LastName.Text = model.LastName.ToString();
                            if (contentItem.Content.UserProfile.ContainsKey("Mobile"))
                                contentItem.Content.UserProfile.Mobile.Text = model.Mobile.ToString();

                            var updateContentContext = new UpdateContentContext(contentItem);
                            user.Properties[def.Name] = JObject.FromObject(contentItem);
                        }
                        _userManager.UpdateAsync(user).GetAwaiter().GetResult();
                    }

**Reply by vip924** [reply]
Hey @yassinsayed ,

Thanks for the info, but when I tried this code, I got an error in the line JObject.FromObject(contentItem)

"JsonSerializationException: Self referencing loop detected for property 'ContentItem' with type 'OrchardCore.ContentManagement.ContentItem'."

**Comment by yassinsayed** [comment]
the below code to retrieve the custom user setting 

            var def = contentDefinitionManager.GetTypeDefinition("UserProfile");
            var contentItem = user.Properties.TryGetValue(def.Name, out JToken property);
            if (contentItem)
            {
                var existing = property.ToObject<ContentItem>();

                if (existing.Content.UserProfile.ContainsKey("FirstName"))
                    FirstName = existing.Content.UserProfile.FirstName.Text;
                
                if (existing.Content.UserProfile.ContainsKey("MiddleName"))                
                    MiddleName = existing.Content.UserProfile.MiddleName.Text;
                
                if (existing.Content.UserProfile.ContainsKey("LastName"))                
                    LastName = existing.Content.UserProfile.LastName.Text;
                
                if (existing.Content.UserProfile.ContainsKey("Mobile"))                
                   Mobile = existing.Content.UserProfile.Mobile.Text;       
            }
